### ISSUE TITLE
Follow the JAX or Tensorflow convention for meaning of grad() with complex inputs

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#41857]
cc #755 

JAX and Tensorflow disagree about whether or not the grad of a complex tensor should be conjugated or not. Here is an easy way to see the difference:

```
from jax import grad

def f(z):
  return z * z

z = 0.1j

print(grad(f, holomorphic=True)(z))
```

gives 0.2j

However

```
>>> x = tf.Variable(0. + 0.j)
>>> sess.run(tf.gradients(x*x, x), feed_dict={x:0.1j})
[-0.20000000000000001j]
```

source: https://github.com/tensorflow/tensorflow/issues/3348

PyTorch also has to decide which side of the field it will come down on. Right now on master it implements JAX. From reading the issue, here is my understanding of the pros and cons:

In favor of TF:
* The gradient is the correct direction for doing gradient descent. This means you can use a "stock" optimizer (one that was written with only real parameters in mind) without any changes to the optimizer. In contrast, to do gradient descent with the JAX definition, you have to remember to conjugate first.

In favor of JAX:
* When the TF definition is implemented directly in the gradient formulas (as opposed to doing a single post facto conjugation), you end up with ugly gradient formulas. For example, take a look at https://github.com/tensorflow/tensorflow/blob/70fd0a4436e3b49139653dc5b85d1c7df23f403d/tensorflow/python/ops/math_grad.py#L453 where TF has to explicitly conjugate the input. With the JAX style definition, you can mostly reuse your real gradient formulas.
* The TF definition is less efficient, since you're doing extra conjugations in the gradient formulas. Even if you write the derivative formulas in the JAX way and then conjugate before plopping the gradient in x.grad, you lose the opportunity to do a fused conjugate-add for the optimizer update.

Gradient formulas look ugly seems like a clear reason to prefer JAX style. Posting this for other opinions.

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen @anjali411 @dylanbespalko @vincentqb

**Comment by vincentqb** [comment]
(just sharing a [comment in another library](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/23#issuecomment-527506584) that adds a little bit to [the tensorflow thread mentioned above](https://github.com/tensorflow/tensorflow/issues/3348))

**Comment by anjali411** [comment]
For holomorphic functions, TF returns the conjugate of gradient defined using Cauchy Riemann

**Comment by boeddeker** [comment]
The following is just my opinion:

I know both definitions under the Wirtinger calculus [1].
Sometimes they are called R-Derivative and Conjugate R-Derivative [2].
In general both are valid and should work.

Let me start with homomorphic function and why this theory cannot be used:
The theory with homomorphic function requires that the real partial derivatives satisfy the Cauchy-Riemann equations:

![image](https://wikimedia.org/api/rest_v1/media/math/render/svg/b2b7e5ead4b2cdcc7ccd78602fc988fc371d56ef)

When you are interested in an optimization problem, you have a real valued objective.
It turns out, that no interesting objective can fulfill the Cauchy-Riemann equations (In general we are interested of the partial derivative of the objective w.r.t. the parameters).
So the theory with homomorphic function cannot be used for optimization and most people use therefore the Wirtinger calculus.

I personally prefer the "TF" style.
 - It allows to interpret the complex number as a vector with two components. The gradient will then be exactly the same.
 - When you work with objectives that contain vectors and matrices and manually calculate the gradient, the "TF" style feels more natural because in most cases you can just replace transpose with transpose conjugate.
 - Regarding the efficiency: When you expose the conjugate argument from the low level functions, I think it will be the same. 

[1]: Brandwood, D. H. "A complex gradient operator and its application in adaptive array theory." IEE Proceedings H-Microwaves, Optics and Antennas. Vol. 130. No. 1. IET, 1983.
[2]: https://arxiv.org/pdf/0906.4835.pdf

**Comment by ezyang** [comment]
> Regarding the efficiency: When you expose the conjugate argument from the low level functions, I think it will be the same.

This is true, although PyTorch internally is not well set up to do this since most of our stuff that isn't going through BLAS doesn't have conjugate arguments (yet, anyway).

@boeddeker, one statement of fact I'm interested in about: is it true that in TF-style you have to do ugly gradient formulas (with extra conjugations)? I came to this conclusion after skimming TF's derivative definitions, but it's possible I drew the wrong conclusion here, and I'd love some confirmation if this is true or false.

**Comment by boeddeker** [comment]
> This is true, although PyTorch internally is not well set up to do this since most of our stuff that isn't going through BLAS doesn't have conjugate arguments (yet, anyway).

I expected an answer like this. Not many high level libraries expose these arguments.

> @boeddeker, one statement of fact I'm interested in about: is it true that in TF-style you have to do ugly gradient formulas (with extra conjugations)? I came to this conclusion after skimming TF's derivative definitions, but it's possible I drew the wrong conclusion here, and I'd love some confirmation if this is true or false.

In [1] we investigated complex valued operations inside an NN.
We used at that time the "TF" style to implement it in chainer, but in the publication we listed both definition.
In Table 4.1, 4.2 and 4.3 you can find the equations for both gradient definitions.
When you have a multiplication you get the conjugate inside the gradient with the "TF" style.
So yes, some gradients contain an extra conjugate.

I personally don't find it ugly. When you work with complex vector and matrix operations it feels more natural, because in many cases you replace transpose with conjugate transpose. This nice fact is kept for the gradient.

[1] https://arxiv.org/abs/1701.00392

**Comment by albanD** [comment]
I think there are few levels of arguments here:
From the user point of view I see three main use cases:
- C -> C functions: there is no easy definition. So whichever we choose should be properly documented and will be good.
- C -> R functions: training complex parameters wrt to a real loss value. Having to  conjugate before the optimizer or not does not sound like a big issue (it should be clearly documented of course). But the provided optimizer can be aware of that and do the right thing out of the box.
- R -> R functions (that happen to contain complex ops in the middle): training a model with real parameters and a real loss that happen to work with complex states. In this case I think that it is important that we return the correct gradient (according to the real definition). This one is more crucial as people will expect this "to just work".

From the dev point of view I do agree that simpler formulas are better. But we will have to audit all of them anyway.
Note that if we start doing `.conj()` for every formula, we might need to add a fast-path for real Tensors here to avoid large perf impact.

**Comment by vincentqb** [comment]
> * C -> C functions: there is no easy definition. So whichever we choose should be properly documented and will be good.

Based on the [comment](https://github.com/pytorch/pytorch/issues/41857#issuecomment-662652780) above by @boeddeker, extending the definition of the complex derivative for holomorphic functions to non-holomorphic functions by going for the Wirtinger derivatives seems to be what users would expect. Are there other extensions of the complex derivatives on holomorphic functions you wanted to consider? Do I understand your point correctly?

**Comment by albanD** [comment]
Yes Wirtinger definition is natural here I think.
From what I understand, the difference between jax and tensorflow in this case is just a conjugate. And you can either have it built-in the autograd or in the optimizer. But maybe I'm missing something.
The one argument I could see for jax here though is that it matches the complex derivative for holomorphic functions.

**Comment by ezyang** [comment]
> I expected an answer like this. Not many high level libraries expose these arguments.

This is a little off topic, but I suggest to @anjali411 that maybe we should support a "conjugate" flag on tensors, which means to interpret the tensor as its conjugate. Then you can very naturally write `matmul(x.conj(), y)` and that will be as efficient as `matmul(x, y, x_conj=True)`

**Comment by ezyang** [comment]
> In [1] we investigated complex valued operations inside an NN.

@boeddeker Thanks, this is a really nice reference. I saw in the paper that you manually calculated out the derivatives; it's gonna be a lot of work to fix up all of our derivative definitions to be this :>

**Comment by boeddeker** [comment]
> From the dev point of view I do agree that simpler formulas are better. But we will have to audit all of them anyway.
> Note that if we start doing .conj() for every formula, we might need to add a fast-path for real Tensors here to avoid large perf impact.

I think this would be a nice feature, independent of the gradient definition.

> Based on the comment above by @boeddeker, extending the definition of the complex derivative for holomorphic functions to non-holomorphic functions by going for the Wirtinger derivatives seems to be what users would expect. Are there other extensions of the complex derivatives on holomorphic functions you wanted to consider? Do I understand your point correctly?

I am not sure if Wirtinger derivative can be interpreted as extension of complex derivative for homomorphic functions.
I would more interpret it as alternative with different properties. I am not aware of an alternative.

> Yes Wirtinger definition is natural here I think.
> From what I understand, the difference between jax and tensorflow in this case is just a conjugate. And you can either have it built-in the autograd or in the optimizer. But maybe I'm missing something.
> The one argument I could see for jax here though is that it matches the complex derivative for holomorphic functions.

Yes, you are right. The difference between both definitions is just the conjugate and you can change between the definitions with the conjugate operation, e.g.: `conj(JAX_gradient) == tensorflow_gradient`.

I found now also one case where JAX definitions has a strange definition: When using the imag operation, you change the sign of the gradient and place the output gradient on the imag part of the input gradient.

> This is a little off topic, but in the past I have suggested to @anjali411 that maybe we should support a "conjugate" flag on tensors, which means to interpret the tensor as its conjugate. Then you can very naturally write matmul(x.conj(), y) and that will be as efficient as matmul(x, y, x_conj=True)

I am not sure, if it is only off topic. It may make the difference if the conjugate operation makes JAX or TF cheaper in terms of computation power. I haven't thought about the intermediate steps, but here an example with TF gradient calculation:
Let's say we have the cost function `w^H R w` (w ... parameter vector, R ... constant hermitian matrix (e.g. cov matrix), H conjugate transpose).
This is a very easy cost function and the TF gradient is simply `2 R w`. So implementing this as a single operation is cheaper than the JAX style (I think w would than get a conj) and when you calculate the intermediate gradients maybe conjugate is called to times on a tensor (I haven't checked this). At least this is an example, where the TF gradient is nicer.

> @boeddeker Thanks, this is a really nice reference. I saw in the paper that you manually calculated out the derivatives; it's gonna be a lot of work to fix up all of our derivative definitions to be this :>

Yes, it was some work. But we learned a lot from it, how NN frameworks work.
And in chainer it was relative easy to integrate.

Initially we started with the JAX definition, but rewrote later everything to the TF style.
Sadly, I forgot why we changed it.
Btw. when you need some help to implement the gradients, I can help.
I have most of them implemented in our teaching NN framework.

**Comment by albanD** [comment]
@boeddeker thanks a lot for taking the time to answer all our questions, it is really helpful.
I would have a couple questions about the paper you linked.

Do I understand correctly that all the derivations that you do (in section 3) are based on the assumption that the overall function is C -> R? (allowing you to pass a single value between layers and recovering the second Wirtinger derivative by conjugation).

What would happen if we apply blindly these formulas for an overall C -> C holomorphic function with all elementary functions being holomorphic (assuming we backpropagate an initial value of `1`)?

What would happen if we apply blindly these formulas for an overall R -> R function that uses complex numbers in the middle of the computations? Can we expect to get the real (as in real space) derivative?

**Comment by boeddeker** [comment]
I am happy to help. I will try to answer your questions.

> Do I understand correctly that all the derivations that you do (in section 3) are based on the assumption that the overall function is C -> R? (allowing you to pass a single value between layers and recovering the second Wirtinger derivative by conjugation).

Yes, the core assumption is, that we only care about C -> R.
When we want to optimize something, we can only do it for a real valued output.
Because of that we rapidly dropped the thought about C -> C and the theory about homomorphic functions.

Since the Wirtinger derivatives are defined as (first is TF style, second JAX)
![image](https://user-images.githubusercontent.com/13744128/88327839-ccb42c80-cd27-11ea-9fef-255659444e48.png)
it is easy to see, that we can obtain the other definition by a conjugation.

> What would happen if we apply blindly these formulas for an overall C -> C holomorphic function with all elementary functions being holomorphic (assuming we backpropagate an initial value of 1)?

I have never thought about this. Maybe it is the same solution as for homomorphic functions.

> What would happen if we apply blindly these formulas for an overall R -> R function that uses complex numbers in the middle of the computations? Can we expect to get the real (as in real space) derivative?

This is a very good question. This reminds me on an internal discussion.
You can apply these definitions to such an operation, but in the backward calculation you have to be careful:
```python
if np.isrealobj(input_argument.value):
    gradient = gradient.real
```
When the input is real valued, but the gradient is complex (i.e. in General the imag part will be unequal to 0) you have to drop the imag part. So when the backprop code handles this detail, you can apply complex valued derivatives to a mixture of real and complex valued tensors.

Just for completeness, the following is a backprop implementation where this is unsed:
<details><summary>backprop code</summary>
<p>
The following code in an extension of the real values backprop from https://github.com/fgnt/2019_ad_xidian/blob/master/practice/ad_solution.ipynb

(Sorry, the code is not very pretty, I added multi output operations as quick fix for the eigenvalue decomposition)

```python
def backprop(loss, out_grad=1., profile=False):
    assert loss.value.shape == np.shape(out_grad), 'The loss should be a scalar. shape {} != {}'.format(
        loss.value.shape,
        np.shape(out_grad),
    )

    map_operation_to_gradient = OrderedDict()
    for operation in reversed(get_operations(loss)):
        if isinstance(operation, MultiOutOperation):
            map_operation_to_gradient[operation] = (0,) * len(operation.value)
        else:
            map_operation_to_gradient[operation] = 0
    map_operation_to_gradient[loss] = out_grad

    if profile:
        times = OrderedDict()

    for operation in map_operation_to_gradient.keys():
        if profile:
            with Timer() as t:
                gradients = operation.grad(map_operation_to_gradient[operation])
            times[operation] = t.time
        else:
            if isinstance(operation, MultiOutOperation):
                gradients = operation.grad(*map_operation_to_gradient[operation])
            else:
                gradients = operation.grad(map_operation_to_gradient[operation])

        assert isinstance(gradients, tuple), (operation, type(gradients), gradients)

        for input_argument, gradient in zip(operation.args + tuple(operation.kwargs.values()), gradients):
            if isinstance(input_argument, Operation):
                if isinstance(input_argument, MultiOutOperation):
                    pass
                else:
                    gradient = np.array(gradient)

                input_operation = input_argument
                gradient = handle_broadcast(
                    input_operation=input_operation,
                    gradient=gradient,
                    current_operation=operation,
                )

                if np.isrealobj(input_argument.value):
                    gradient = gradient.real

                if isinstance(input_argument, MultiOutOperation):
                    map_operation_to_gradient[input_operation] = tuple(
                        map(
                            operator.add,
                            map_operation_to_gradient[input_operation],
                            gradient,
                        )
                    )
                else:
                    map_operation_to_gradient[input_operation] = (
                            map_operation_to_gradient[input_operation]
                            + gradient
                    )
            # else:
            #     raise ValueError(input_argument, gradient)

    if profile:
        ret = OrderedDict()
        for k in map_operation_to_gradient.keys():
            ret[k] = {'grad': map_operation_to_gradient[k], 'time': times[k]}
        return ret
    else:
        return map_operation_to_gradient
```

</p>
</details>

**Comment by ezyang** [comment]
On Friday, @anjali411 @vincentqb @albanD and I met and we went through the derivation from Wirtinger derivatives to get the JAX formula and the TF formula. But we didn't come to an agreement about what formulation we should use. @anjali411 and @albanD thought that JAX-style would be better (primarily citing reducing the amount of work we would have to do in the framework); @vincentqb thought TF-style was better but didn't hold the opinion too strongly, and I thought that we should switch to TF (citing that we are primarily an optimization library and therefore we should use the definition that's appropriate for optimization).

**Comment by SsnL** [comment]
> we are primarily an optimization library and therefore we should use the definition that's appropriate for optimization

agreed!

**Comment by ngimel** [comment]
I also agree that we should use definition appropriate for optimization. Also, in terms of implementation, like @ezyang suggests, tensors should have "conj" flag, so the formulas can remain the same, and we could just set the flag for the `.grad`. There should be some API to materialize a tensor with the `conj` flag, similar to how now we have `.t().contiguous()`.

**Comment by ezyang** [comment]
One idea that I had for implementing a conj flag is to implement it as a dispatch key with backend fallback. Tensors with the conjugate bit set would also carry the Conjugate dispatch key. The fallback handler would materialize the conjugation and then call the underlying operation; however, specific operations which support conjugations on certain inputs could simply strip the conjugate bit and forward to an underlying operation that supports explicitly handling the conjugation.

**Comment by anjali411** [comment]
I think we should follow the tensor as well, since the application of interest for majority of complex users is being able to optimize with real valued objective and real/complex parameters. Switching to tensorflow convention would require addition of conj in some ops, for e.g., `torch.mul`, `torch.div`, torch.abs`, `torch.matmul`, `torch.inverse` etc.. The vjp definition for some other ops for e.g., `torch.add`, `torch.conj`, `torch.real` would remain unaffected. It seems like most (if not all) of the backward definitions for linear algebra ops will need an extra conj because of the `Hermitian` operator
.
@boeddeker I found Section 4 of https://arxiv.org/pdf/1701.00392.pdf to be a very helpful resource. thank you!

Since we will need to add a bunch of conj for some operations, I was wondering if having an inplace conj `torch.conj_` instead of `torch.conj` in the vjp definitions would resolve the issue since it would be a no op for real tensors. Ed had some concerns about this approach specifically that it might break higher order derivatives since it would mutate intermediates that are used for the double backwards. cc. @albanD will that be the case?

**Comment by ezyang** [comment]
Amusingly enough: https://github.com/google/jax/issues/4891

**Comment by mattjj** [comment]
That JAX issue was spurred by of a conversation at the PyTorch dev meeting!

There's no "right" answer here, as it's just a convention for the `grad` wrapper, but I think it's slightly cleaner to define `grad` as `jnp.vdot(grad(f)(x), v) == jvp(f, (x,), (v,))` rather than `jnp.dot(grad(f)(x), v) == jvp(f, (x,), (v,))` (where `vdot` models the standard inner product, where `dot` models matrix-vector multiplication). Luckily `jvp` and `vjp` are never ambiguous!

**Comment by mattjj** [comment]
By the way, just because I don't see it mentioned here and because I think it provides a useful overview of these ideas, it might be worth checking out the [JAX autodiff cookbook section on differentiation with complex numbers](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation).

**Comment by anjali411** [comment]
closing this issue now that we have switched to TF style gradients from 1.7 release.
